genesis/streaming.py

```diff
@@ -14,350 +14,414 @@
-			if not is_heartbeat(msgs):
-				self._buffer = dict( enumerate(msgs) )
-
-				# process the messages on the workers
-				for i, (rec, meta) in enumerate(mapper(ParseAndFilter(parser=self._parser, filter=filter), msgs)):
-					# pop the message from the buffer, indicating we've processed it
-					del self._buffer[i]
-
-					# mark as consumed and increment the index _before_ we yield,
-					# as we may never come back from yielding (if the user decides
-					# to break from the loop)
-					self._consumed[(meta.topic, meta.partition)] = meta.offset
-
-					# yield if the filter didn't return None
-					if rec is not None:
-						yield (rec, meta)
-
-	def commit(self, defer=True):
-		self._committed = self._consumed.copy()
-		self._commit_to_kafka(defer=defer)
-
-	def _commit_to_kafka(self, defer=True):
-		# Occasionally commit read offsets (note: while this ensures we won't lose any
-		# messages, some may be duplicated if the program is killed before offsets are committed).
-		now = time.time()
-		last = self.last_commit_time if self.last_commit_time is not None else 0
-		if self._committed and (not defer or now - last > 5.):
-			print("COMMITTING", file=sys.stderr)
-			tp = [ TopicPartition(_topic, _part, _offs + 1) for (_topic, _part), _offs in self._committed.items() ]
-			#print("   TP=", tp)
-			self.c.commit(offsets=tp)
-
-			# drop all _committd offsets from _consumed; no need to commit them again if the user
-			# calls commit()
-			self._consumed = {k:v for k,v in self._consumed.items() if k not in self._committed}
-			self._committed = {}
-
-			self.last_commit_time = now
-
-	def _stream(self, filter, mapper, progress, timeout, limit):
-		import warnings
-		with warnings.catch_warnings():
-			# hide the annoying 'TqdmExperimentalWarning' warning
-			warnings.simplefilter("ignore")
-			from tqdm.autonotebook import tqdm
-
-		t = tqdm(disable=not progress, total=limit, desc='Alerts processed', unit=' alerts', mininterval=0.5, smoothing=0.5, miniters=0)
-
-		nread = 0
-		for rec, meta in self._filtered_stream(mapper=mapper, filter=filter, timeout=timeout):
-			yield (rec, meta) if self._metadata else rec
-
-			t.update()
-			nread += 1
-
-			if nread == limit:
-				break
-
-		t.close()
-
-	# returns a generator for the user-mapped messages using the filter function,
-	# possibly executed on ncores and up to maxread values
-	def __call__(self, filter=None, pool=None, progress=False, timeout=None, limit=None):
-		if pool:
-			mapper = lambda fun, vec: pool.imap(fun, vec, chunksize=100)
-		else:
-			mapper = map
-
-		yield from self._stream(filter, mapper=mapper, progress=progress, timeout=timeout, limit=limit)
-
-	def __iter__(self):
-		return self.__call__()
-
-	# producer functionality
-	def write(self, msg):
-		packet = self._serialize(msg)
-		self.p.produce(self.topics[0], packet)
-
-	def flush(self):
-		return self.p.flush()
+    c = None
+    p = None
+
+    def __init__(
```

Comment:
I think that one-argument-per-line style is not very common -- PEP8 recommends this:
```
# Aligned with opening delimiter.
foo = long_function_name(var_one, var_two,
                         var_three, var_four)
```
(https://www.python.org/dev/peps/pep-0008/#indentation), and that jibes intuitively with what I'd expect in when reading a Python file (Python's own standard library seems to follow this too).

Reply:
To defend this a bit, I don't think PEP8 makes any particular recommendations on splitting functions with many arguments, the only guideline is that there's enough indentation that it's clear it's a continuation line.

Personally, I'm not a fan of that style as the leading whitespace doesn't correspond to any real indentation level.

Some reasoning/discussion for the one-line-per-argument for long argument calls is given here: https://stackoverflow.com/a/57814233, and more recently with the typing module here: https://github.com/python/typing/issues/433#issuecomment-302452660

That being said, I'm good just picking a convention and going with it.

Reply:
Yeah, those are all _definitely_ good reasons (esp. with typing), but it's still relatively unusual to see in the wild. I'm no big fan of hanging indent either, but having the code look "natural" to folks in related projects/fields brings us a lot of value.

Let's standardize on the "Aligned with opening delimiter" variant -- that keeps us aligned with LSST (and also astropy).

Reply:
If we still want to reap the benefits of a code formatter but don't want to conform fully to black's opinionated nature, there's also `autopep8` which also can piggyback on the flake8 configuration.

As an example, adding this to setup.cfg:
```
[flake8]
max_line_length = 100
ignore = E501
```

and replacing `black` with `autopep8` in `make format` appears to avoid some of the concerns you were having, specifically the many argument case and the dictionary formatting.

Reply:
If we have all of our code autoformatted, then we can make it a prerequisite for merging. If we do that, it we can eliminate a whole genre of code review back-and-forth, and eliminate a lot of thought in the heads of developers. You don't worry about formatting, you just write code and then run make format and commit what you have.

I don't really care which autoformatter we use. Black is fine, so is autopep8, so is yapf. I just want one standard, _ideally_ unconfigured, which we can use automatically everywhere.

This has worked exceptionally well in the Go community, where [gofmt](https://golang.org/cmd/gofmt/) is law. There's a go proverb that "Gofmt's style is no one's favorite, yet gofmt is everyone's favorite," which really is true - sometimes it does stuff that looks a little odd, but the universal application of it is really wonderful.

I'll go with `autopep8` and maybe make another PR showing it off.

Reply:
Here, you can take a look at [output of `autopep8`](https://github.com/astronomy-commons/genesis-client/compare/enhancement/autopep8) and [of yapf](https://github.com/astronomy-commons/genesis-client/compare/enhancement/yapf).

Reply:
I'll put my vote on autopep8 given the constraints above and it being tied more closely together with flake8.

Reply:
Ah, I didn't realize `black/autopep8` were a autoformatters! Hardcore :).

Adding my +1 to `autopep8` -- the auto-formatted code you linked to looks very pleasent.